File: MasteringDart/lib/chapters/chapter04/intermediate/intermediate_operators.dart
// ignore_for_file: avoid_print
//
//  intermediate_operators.dart
//  MasteringDart
//
//  Capítulo 04 — Operadores y expresiones.
//  Nivel INTERMEDIO: Combinar comparaciones, operadores lógicos y condiciones.
//
//  Ejecutar con: dart run lib/chapters/chapter04/intermediate/intermediate_operators.dart

/// Ejecuta los ejemplos intermedios de operadores.
void runIntermediateOperatorExamples() {
  // --- Validación de edad (operadores de comparación: <, <=, >, >=, ==, !=) ---
  const age = 17;
  const minAge = 18;
  const maxAge = 120;
  final isAdult = age >= minAge;
  final isValidAge = age >= 0 && age <= maxAge;
  print(
      'Age: $age — isAdult (>= $minAge): $isAdult, isValid (0..$maxAge): $isValidAge');

  final ageMessage = age >= minAge
      ? 'Access granted.'
      : 'You must be $minAge or older.';
  print(ageMessage);

  // --- Comparación de números ---
  const num1 = 10;
  const num2 = 20;
  print(
      'Comparison: $num1 == $num2 -> ${num1 == num2}, $num1 < $num2 -> ${num1 < num2}, $num1 != $num2 -> ${num1 != num2}');

  // --- Operadores lógicos (&&, ||, !) ---
  const hasTicket = true;
  const hasId = true;
  const isBanned = false;
  final canEnter = hasTicket && hasId && !isBanned;
  print('Logical: hasTicket && hasId && !isBanned -> canEnter = $canEnter');

  const score = 85;
  const passingScore = 60;
  const bonusThreshold = 90;
  final passed = score >= passingScore;
  final gotBonus = score >= bonusThreshold;
  final passedOrBonus = passed || gotBonus;
  print(
      'Score $score: passed=$passed, gotBonus=$gotBonus, passedOrBonus=$passedOrBonus');

  // --- Expresiones booleanas con precedencia de operadores ---
  const temp = 25;
  const isSunny = true;
  final isWarmDay = temp > 20 && isSunny;
  final isOkDay = temp >= 15 || isSunny;
  print(
      'temp=$temp, isSunny=$isSunny -> isWarmDay=$isWarmDay, isOkDay=$isOkDay');
}

void main() {
  runIntermediateOperatorExamples();
}

File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter04/intermediate/IntermediateOperators.kt
package com.devlokos.masteringkotlin.chapters.chapter04.intermediate

/**
 * Capítulo 04 — Operadores y expresiones.
 * Nivel INTERMEDIO: Combinar comparaciones, operadores lógicos y condiciones.
 *
 * Ejemplos: validación de edad, comparación de números, expresiones booleanas.
 */
fun main() {
    runIntermediateOperatorExamples()
}

fun runIntermediateOperatorExamples() {
    // --- Validación de edad (operadores de comparación: <, <=, >, >=, ==, !=) ---
    val age = 17
    val minAge = 18
    val maxAge = 120
    val isAdult = age >= minAge
    val isValidAge = age in 0..maxAge
    println("Age: $age — isAdult (>= $minAge): $isAdult, isValid (0..$maxAge): $isValidAge")

    val ageMessage = if (age >= minAge) "Access granted." else "You must be $minAge or older."
    println(ageMessage)

    // --- Comparación de números ---
    val num1 = 10
    val num2 = 20
    println("Comparison: $num1 == $num2 -> ${num1 == num2}, $num1 < $num2 -> ${num1 < num2}, $num1 != $num2 -> ${num1 != num2}")

    // --- Operadores lógicos (&&, ||, !) ---
    val hasTicket = true
    val hasId = true
    val isBanned = false
    val canEnter = hasTicket && hasId && !isBanned
    println("Logical: hasTicket && hasId && !isBanned -> canEnter = $canEnter")

    val score = 85
    val passingScore = 60
    val bonusThreshold = 90
    val passed = score >= passingScore
    val gotBonus = score >= bonusThreshold
    val passedOrBonus = passed || gotBonus
    println("Score $score: passed=$passed, gotBonus=$gotBonus, passedOrBonus=$passedOrBonus")

    // --- Expresiones booleanas con precedencia de operadores ---
    // Precedencia: comparación antes que && antes que ||
    val temp = 25
    val isSunny = true
    val isWarmDay = temp > 20 && isSunny
    val isOkDay = temp >= 15 || isSunny
    println("temp=$temp, isSunny=$isSunny -> isWarmDay=$isWarmDay, isOkDay=$isOkDay")
}
